File: old/Stack/C#/Stack.cs
namespace C_
{
    public class MyStack<T>
    {
        public Node<T>? Top { get; private set; }
        public int Length
        {
            get
            {
                var currentNode = Top;
                var currentIndex = 0;

                while (currentNode != null)
                {
                    currentNode = currentNode.Next;
                    ++currentIndex;
                }

                return currentIndex;
            }
        }
        public Node<T> this[int index]
        {
            get
            {
                if (index < 0) throw new IndexOutOfRangeException();

                var currentNode = Top;
                var currentIndex = 0;

                while (currentNode != null)
                {
                    if (currentIndex++ == index) return currentNode;
                    currentNode = currentNode.Next;
                }

                throw new IndexOutOfRangeException();
            }
        }
        public void Push(T value)
        {
            var newNode = new Node<T>(value);

            if (Top == null)
            {
                Top = newNode;
                return;
            }

            newNode.Next = Top;
            Top = newNode;
        }
        public T? Pop()
        {
            if (Top != null)
            {
                var value = Top.Value;
                Top = Top.Next;
                return value;
            }

            return default;
        }
    }
}

File: old/Stack/C++/src/Stack.cpp
#ifndef STACK_CPP
#define STACK_CPP

#include "stack.h"

template <typename T>
Stack<T>::~Stack()
{
    Node<T>* current = top;

    while (current != nullptr) {
        Node<T>* next_node = current->next;
        delete current;
        current = next_node;
    }
}

template <typename T>
void Stack<T>::push(const T value)
{
    Node<T>* new_node = new Node<T>(value);

    if (top != nullptr)
        new_node->next = top;

    top = new_node;
}

template <typename T>
T Stack<T>::pop()
{
    if (top != nullptr)
    {
        T value = top->value;
        top = top->next;
        return value;
    }

    return T(); // return default type value
}

template <typename T>
T* Stack<T>::toArray(size_t& size) const
{
    size = 0;
    Node<T>* current = top;

    while (current != nullptr)
    {
        size++;
        current = current->next;
    }

    T* values = new T[size];
    current = top;
    int index = 0;

    while (current != nullptr)
    {
        values[index] = current->value;
        index++;
        current = current->next;
    }

    return values;
}

#endif
